### PR TITLE
Fix remaining numpy mypy issues

### LINKED ISSUE #347: Numpy-related type errors are hidden by the absence of numpy stubs
**Describe the bug**
There is a bunch of type errors in the code base that mypy runs never reveal because Trieste is relying on numpy 1.19.x which isn't typed yet. If you run mypy with numpy >= 1.20, 46 new type errors surface.

Ideal fix would be to move to newer version of numpy. The errors will then fail the builds and force us to fix them. Unfortunately numpy 1.19.x seems to be forced by [Tensorflow](https://github.com/tensorflow/tensorflow/blob/07dd16b3a09e0cfa007d9694251dd6b2cbdf36ad/tensorflow/lite/tools/pip_package/setup_with_binary.py) and using one numpy version for type checks while using older one for everything else feels weird.

We could also consider installing numpy stubs separately via [numpy-stubs](https://github.com/numpy/numpy-stubs) or [data-science-types](https://github.com/predictive-analytics-lab/data-science-types), but both of these projects are archived, so it's probably best not to rely on them.

**To reproduce**
We will be using tox here.

1. Start with a new fresh clone of Trieste, make sure there is no `.tox` cache. Your venv should be running python3.7 and only have tox installed. This is how GitHub Action for type checking is setup.
2. Add numpy and matplotlib to tox requirements:
`echo -e "numpy>=1.20\nmatplotlib" >> common_build/types/requirements.txt`
3. run type checks with tox
`tox -e types`
4. 46 errors show up:  [mypy_errors.txt](https://github.com/secondmind-labs/trieste/files/7153689/mypy_errors.txt)

**Expected behaviour**
No errors

**System information**
 - OS: Ubuntu 18
 - Python version: 3.7.5
 - Trieste version: git clone

### PR DESCRIPTION
**Pull request by uri-granta**
This fixes the remaining issues exposed by the numpy stubs introduce in 1.21 and adds a workflow to test this.

### CODE REVIEW

File: tests/unit/acquisition/test_utils.py [hunk 1]
```diff
@@ -41,7 +42,7 @@
     ],
 )
 def test_split_acquisition_function(
-    f: AcquisitionFunction, x: np.ndarray, split_size: int, expected_batches: int
+    f: AcquisitionFunction, x: np.ndarray[Any, Any], split_size: int, expected_batches: int
```
Comment (hstojic) [inline tests/unit/acquisition/test_utils.py]: so here we explicitly state the type of objects in arrays, similar as with tuple, List etc?

**Review by hstojic — APPROVED** [review]
looks good to me

File: tests/unit/acquisition/test_utils.py (hunk 1 above)
Comment (uri-granta) [reply tests/unit/acquisition/test_utils.py]: Yes: the parameters refer to shape and dtype (though shape isn't properly supported at the moment, and we currently don't handle dtypes well enough to use that part). There's a cleaner shorthand in numpy.typing called `npt.NDArray` but that doens't work with older versions of numpy.

File: tests/unit/models/conftest.py [hunk 2]
```diff
@@ -95,9 +95,8 @@ def _two_layer_model_fixture(request: Any) -> Callable[[TensorType], DeepGP]:
 
 
 # Teardown fixture to set keras floatx to float64 then return it to previous value at test finish
-# pytest uses yield in a funny way, so we use type ignore
-@pytest.fixture(name="keras_float")  # type: ignore
-def _keras_float() -> None:
+@pytest.fixture(name="keras_float")
```
Comment (hstojic) [inline tests/unit/models/conftest.py:98]: should we then move this fixture up to `tests/conftest.py` since we have it now in multiple places?

Comment (uri-granta) [reply tests/unit/models/conftest.py:98]: :+1: but planning to tidy up the conftests in a separate PR

**Pull request merged**